Add a control per block to reset pattern overrides

kevin940726 opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
Part of #53705 and https://github.com/WordPress/gutenberg/issues/57751. Continued from https://github.com/WordPress/gutenberg/pull/57845.

Add a control per block to reset pattern overrides.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

As mentioned in https://github.com/WordPress/gutenberg/issues/57751.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

Add a control and conditionally render it.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->
Follow the instructions in https://github.com/WordPress/gutenberg/issues/53705#issuecomment-1882305331 to create a pattern with overrides and add it to a post.

1. Focus on one of the overridable blocks, and it should have a disabled control "Reset to original" in the toolbar.
2. Make some changes to the block, and the control should be enabled.
3. Clicking on it should reset only that block to the original state.
4. Undo/redo should work as expected.

## Screenshots or screencast <!-- if applicable -->
Block without overrides
![Block without overrides](https://github.com/WordPress/gutenberg/assets/7753001/e3a26f95-5ddc-4d3d-ab7d-e4ee8ca7a843)

Block with overrides
![Block with overrides](https://github.com/WordPress/gutenberg/assets/7753001/3a095f81-eb2e-4101-ac6a-a07020b3ab69)

kevin940726 committed:
Add a control per block to reset pattern overrides

kevin940726 commented on packages/block-library/src/block/edit.js:
```diff
@@ -117,6 +118,16 @@ function applyInitialOverrides( blocks, overrides = {}, defaultValues ) {
+		attribute1 instanceof RichTextData &&
+		attribute2 instanceof RichTextData
+	) {
+		return attribute1.toString() === attribute2.toString();
```
A bit of a gotcha here that if the attributes are different `RichTextData` instances then `===` won't be true even though we expect them to be most of the time. 🤔

talldan replied:
I can't really think of another way, so this looks ok to me.

kevin940726 committed:
Reword the control

talldan committed:
Update code to check for new block condensed bindings format

talldan reviewed (commented):
At the moment it's not quite working for me, the reset button always appears greyed out even after making edits:
![Screenshot 2024-02-06 at 1 45 44 pm](https://github.com/WordPress/gutenberg/assets/677833/86e770ed-fbde-4de8-84d2-7bb1b433ca57)

I rebased and pushed a commit for an issue where the button wasn't showing at all (code needed to be updated for the new block bindings format), so it could be something similar.

I think it'd be good if the button could be moved to the end of the toolbar rather than the start, it's quite prominent at the moment.

kevin940726 committed:
Fix the name field

kevin940726 committed:
Add e2e test

kevin940726 commented:
Just fixed the issue and moved the reset button to the end.

![image](https://github.com/WordPress/gutenberg/assets/7753001/bb69004c-d2da-400f-adfd-de7586f36b88)

I also added an e2e test.

talldan commented on packages/patterns/src/components/reset-overrides-control.js:
```diff
@@ -0,0 +1,75 @@
+import { __ } from '@wordpress/i18n';
+
+function recursivelyFindBlockWithId( blocks, id ) {
+	return blocks.find( ( block ) => {
+		if ( block.attributes.metadata?.id === id ) {
+			return block;
+		}
+
+		return recursivelyFindBlockWithId( block.innerBlocks, id );
+	} );
```
@kevin940726 I don't think this does what's expected. `blocks.find` will only ever return a top-level block since it gets the value from the current array that the callback returns `truthy` for.

If you try a pattern where an overriden paragraph is nested in a group it doesn't work as expected.

kevin940726 replied:
Oops! I don't know what I was thinking 😅. Fixed in [6ec77d7](https://github.com/WordPress/gutenberg/pull/57907/commits/6ec77d73346f147709e6ab434f8bb4335e4fcdb2). Hopefully I'm not missing something obvious again! 🤞

kevin940726 committed:
Fix nested block

glendaviesnz reviewed (approved):
This is testing well for me now, including with deeply nested overridden blocks
